### ISSUE TITLE
[DOC] Document that certain objects require manual cleanup

### ISSUE THREAD

**Issue by leofang** [mpi4py/mpi4py#32]
For example, users need to call `Request.Free()` explicitly, see the discussion around https://github.com/mpi4py/mpi4py/pull/29#discussion_r593384270.

**Comment by leofang** [comment]
This needs:
1. a short example added to Tutorial
2. docstrings added to the relevant classes (ex: `Request`)

**Comment by leofang** [comment]
@dalcinl If IIUC non-default instances of the following 9 (*) classes need an explicit destructor call:
- Datatype
- Win
- Request
- Op
- Info
- Group
- Comm
- Errhandler
- File (* should call `Close()` instead of `Free()`)

**Comment by dalcinl** [comment]
Yes. `Request` is a bit special. For `Request` instances, `Wait()` or `Test()` methods implicitly work as a destructor. `Prequest` (subclass of `Request`) instances always require a `Free()` call.

The case of `Comm`, `Win`, and `File` is also a bit special, as the `Free()` or `Close()` calls are collective. A consequence of this is that these destructors could never be let in the hand of Python's garbage collector.

**Comment by FreddieWitherden** [comment]
For the `Prequest` type is there scope to adding a method/flag which says "please call Free for me when my Python refcount goes to zero; I'll accept the consequences"?

**Comment by dalcinl** [comment]
> For the `Prequest` type ...

Well, I think that would also apply to almost all types, with the exceptions of `Comm`/`Win`/`File`. 

I acknowledge this is extremely annoying and unpythonic. The primary reason I've always refused to call free automatically on garbage collection is C/C++/Fortran interoperability. There is no obvious way to take or relinquish ownership of MPI objects. mpi4py has always tried hard to be a thin wrapper, without adding too much "magic" and API bloat to the core MPI specification.

I do have some preliminary, preparatory work for something like what you proposed using the bitset `flags` slot in the Python object structures. But I could not manage to get a simple ownership-handling API that would cover all the use cases related to C/C++/Fortran interoperability. Maybe I did not try hard enough. 

Both MPICH and Open MPI use refcounts internally, but the incref/decref operations are not part of the MPI standard. What a pity.

**Comment by FreddieWitherden** [comment]
I think not calling free automatically is a reasonable default; better to leak the occasional reference than segfault.

For me the simplest ownership API would be something like:

```python
class Prequest:
    def __init__(self, ...):
        ...
        self._gc = False

    def __del__(self):
        if self._gc:
            self.Free()

    def set_gc(self, state):
        self._gc = state
```
where if you opt in you are making a commitment to keep the Python object alive so long as other code might have outstanding references.  This actually isn't *too* bad: in our code we wrap CUDA and HIP and have the exact same issue when it comes to memory allocations (how do we know someone else is not holding onto a pointer or slice thereof).  However, it quickly becomes quite natural to keep references around.  In contrast explicit Free'ing can be very awkward.  You need to code extremely defensively (try/finally) to be exception safe and memoization becomes a pain.  For example, here in our all-to-all wrapper I don't even try to cache the data type:

https://github.com/FreddieWitherden/PyFR/blob/develop/pyfr/mpiutil.py#L109

instead it is just created, used once, and released because of the hassle.  We do a similar thing with operators: recreate and free.

**Comment by dalcinl** [comment]
@FreddieWitherden In the mean time, if you can afford to depend on the mpi4py release 4.0.x, you can take advantage of the improved support for weak references, plus a few other small new APIs/features I've added. Of  course, I do not expect users to know about them, who reads a changelog these days :laughing: 

```python
import weakref
from mpi4py import MPI

def autofree(obj):
    def callfree(fromhandle, handle):
        fromhandle(handle).free()
    weakref.finalize(obj, callfree, obj.fromhandle, obj.handle)
    return obj


comm = autofree(MPI.COMM_SELF.Dup())
del comm

dtype = autofree(MPI.DOUBLE.Dup())
del dtype
```

Note I'm using the lowercase method `free()` for freeing the MPI objects. This method works very much like `Free()`, except that it will ignore any `MPI_XXX_NULL` or built-in object that would make `MPI_XXX_free()` fail. It also uses `handle`/`fromhandle`, where `handle` is basically an integral (`uintptr_t`) representation of the MPI handle which works equally well (in the sense of preserving value) with regards to 32bit int vs. 64bit pointer representations, and is designed to be future-proof regarding the MPI ABI proposal. Of course, none of this would work without mpi4py adding support for weak references and the addition of [`weakref.finalize`](https://docs.python.org/3/library/weakref.html#weakref.finalize) to Python 3.4+. 

If you cannot afford to depend on mpi4py 4.0.x, the code above can be adjusted to work with mpi4py 3.1.x, but it would have to use `Free()` with a check for `NULL` handles, plus relying in the Fortran/C handle conversion routines, like the following (untested):

```python
def autofree(obj):
    def callfree(fromhandle, handle):
        ob = fromhandle(handle)
        if ob: ob.Free()
    weakref.finalize(obj, callfree, obj.f2py, obj.py2f())
    return obj
```

IMHO, the code above is good enough for an opt-in approach. Do you like it?
Maybe this trick would be worth an entry in the documentation, but I'm such a bad documenter...

**Comment by dalcinl** [comment]
> Do all of the major types now support free? We depend on the 4.x series of mpi4py but for some reason I feel like a type or two does not support free and instead require the upper-case version?

I hope it is just a feeling :laughing: ... Can you elaborate?
AFAICT, every (base) type supports the lowercase `free()` method, with the notable exception of Status simply because for it it does not make sense. Also note that for `File`, `free()` will actually invoke `Close()`, but the lowercase method is still named `free()`. Similarly for `Session` and its `Finalize()` method.

> As for the MPI ABI proposals: I have a bottle of Champagne ready for when OpenMPI gets ABI compatibility with everyone else!

:crossed_fingers:

**Comment by zerothi** [comment]
> No, it is not a miss. The `Status` type is a bit special, when a Python `Status` object is created, there is no extra memory allocation for the `MPI_Status` C entry. Moreover, the `MPI_Status` does not hold resources internally, therefore there is no `MPI_Status_free()` function to call. Once Python's GC gets rid of the Python Status, that's all what is needed.

thanks for the clarification!
